docs(lep): v2 sharding (#13159)
longhorn/longhorn: enhancements/20260519-v2-sharding.md

# V2 Sharding

This document provides a detailed proposal and design for the Longhorn V2 sharding feature - erasure coded storage with Reed-Solomon coded data distribution across multiple physical devices.

> 

Reviewer: Suggest adding a TOC

```
## Table of Contents

- [V2 Sharding](#v2-sharding)
  - [Table of Contents](#table-of-contents)
  - [Summary](#summary)
    - [Related Issues](#related-issues)
  - [Reviewer's overview](#reviewers-overview)
    - [Topology in one picture](#topology-in-one-picture)
    - [Six load-bearing design decisions](#six-load-bearing-design-decisions)
    - [Reading paths by role](#reading-paths-by-role)
  - [Motivation](#motivation)
    - [Goals](#goals)
    - [Non-goals](#non-goals)
  - [Proposal](#proposal)
    - [SPDK Proposal 1 - EC Full-Stripe write (7-Layered stack)](#spdk-proposal-1---ec-full-stripe-write-7-layered-stack)
    - [SPDK Proposal 2 - EC Read-Modify-Write (4-Layered stack)](#spdk-proposal-2---ec-read-modify-write-4-layered-stack)
    - [Decision](#decision)
    - [User Stories](#user-stories)
      - [Story 1: Reduced Storage Cost](#story-1-reduced-storage-cost)
      - [Story 2: Online Expansion](#story-2-online-expansion)
      - [Story 3: Disk Failure and Rebuild](#story-3-disk-failure-and-rebuild)
    - [User Experience In Detail](#user-experience-in-detail)
  - [Design](#design)
    - [SPDK Implementation Overview](#spdk-implementation-overview)
      - [Threading model](#threading-model)
      - [Mental model: layout and the three maps](#mental-model-layout-and-the-three-maps)
      - [Async bdev\_ec\_create](#async-bdev_ec_create)
      - [How data is laid out:](#how-data-is-laid-out)
      - [How address translation works through the stack:](#how-address-translation-works-through-the-stack)
      - [How reads work:](#how-reads-work)
      - [How write works](#how-write-works)
        - [Counters](#counters)
      - [Dirty bitmap details](#dirty-bitmap-details)
      - [Crash safety of the dirty bitmap](#crash-safety-of-the-dirty-bitmap)
      - [WIB on-disk layout](#wib-on-disk-layout)
      - [Front-placed metadata: WIB and unmapped bitmap share a layout](#front-placed-metadata-wib-and-unmapped-bitmap-share-a-layout)
      - [WIB region granularity and sizing](#wib-region-granularity-and-sizing)
        - [The region size is fixed at 1024 stripes per bit](#the-region-size-is-fixed-at-1024-stripes-per-bit)
          - [Why 1024?](#why-1024)
        - [Maximum volume size](#maximum-volume-size)
          - [About `strip_size_kb`](#about-strip_size_kb)
          - [How k and m affect WIB sizing](#how-k-and-m-affect-wib-sizing)
          - [How to choose k, m, and strip\_size](#how-to-choose-k-m-and-strip_size)
          - [Recommended configurations by volume size](#recommended-configurations-by-volume-size)
      - [WIB persist lifecycle](#wib-persist-lifecycle)
      - [Startup scrub:](#startup-scrub)
      - [How UNMAP works (capacity reclamation)](#how-unmap-works-capacity-reclamation)
        - [Why a persistent unmapped bitmap](#why-a-persistent-unmapped-bitmap)
        - [On-disk layout of the unmapped bitmap](#on-disk-layout-of-the-unmapped-bitmap)
        - [Setting a stripe unmapped (UNMAP path)](#setting-a-stripe-unmapped-unmap-path)
        - [Split dispatcher (unaligned multi-stripe)](#split-dispatcher-unaligned-multi-stripe)
        - [Reading an unmapped stripe (correctness-critical)](#reading-an-unmapped-stripe-correctness-critical)
        - [Writing into an unmapped stripe (clearing a bit)](#writing-into-an-unmapped-stripe-clearing-a-bit)
        - [Bit-clear waiter queue](#bit-clear-waiter-queue)
        - [Scrub and rebuild skip unmapped stripes (optimization)](#scrub-and-rebuild-skip-unmapped-stripes-optimization)
        - [Concurrency](#concurrency)
        - [Resize](#resize)
        - [Native WRITE\_ZEROES path](#native-write_zeroes-path)
        - [Counters](#counters-1)
      - [IO splitting](#io-splitting)
      - [Why splitting is necessary](#why-splitting-is-necessary)
      - [EC bdev operation states:](#ec-bdev-operation-states)
      - [How disk failure is handled:](#how-disk-failure-is-handled)
      - [NVMe-oF transient failure vs permanent disk failure](#nvme-of-transient-failure-vs-permanent-disk-failure)
      - [Shard node reboot vs permanent disk failure](#shard-node-reboot-vs-permanent-disk-failure)
      - [How disk replacement and rebuild work:](#how-disk-replacement-and-rebuild-work)
      - [How foreground writes and rebuild interact:](#how-foreground-writes-and-rebuild-interact)
      - [How in-place resize works (same k/m, bigger disks):](#how-in-place-resize-works-same-km-bigger-disks)
      - [How snapshot work with the EC stack:](#how-snapshot-work-with-the-ec-stack)
        - [EC snapshot implementation: ShardGroup-local lvol operations (not per-shard, not engine-local)](#ec-snapshot-implementation-shardgroup-local-lvol-operations-not-per-shard-not-engine-local)
      - [How teardown works:](#how-teardown-works)
      - [How crash recovery works (end-to-end):](#how-crash-recovery-works-end-to-end)
    - [Longhorn Data-plane Implementation Overview](#longhorn-data-plane-implementation-overview)
      - [Engine role for EC volumes (clarification)](#engine-role-for-ec-volumes-clarification)
      - [Bdev stack construction:](#bdev-stack-construction)
        - [Engine layout dispatch](#engine-layout-dispatch)
      - [Health event propagation](#health-event-propagation)
      - [Shard status propagation (control-plane \<-\> data-plane)](#shard-status-propagation-control-plane---data-plane)
        - [Two-stage flow: push (event) + pull (state)](#two-stage-flow-push-event--pull-state)
      - [Startup and crash recovery:](#startup-and-crash-recovery)
    - [Longhorn Control-plane Implementation Overview](#longhorn-control-plane-implementation-overview)
      - [CRD layering and responsibilities](#crd-layering-and-responsibilities)
        - [Linkage between CRs](#linkage-between-crs)
      - [New CRD](#new-crd)
      - [Volume controller EC mode detection](#volume-controller-ec-mode-detection)
      - [CRD validation (webhook)](#crd-validation-webhook)
      - [ShardGroup controller architecture](#shardgroup-controller-architecture)
      - [Engine node selection](#engine-node-selection)
      - [Scheduling (shard placement)](#scheduling-shard-placement)
      - [Orphaned shard instance cleanup](#orphaned-shard-instance-cleanup)
      - [Failure recovery (ShardGroup controller reconcile path):](#failure-recovery-shardgroup-controller-reconcile-path)
      - [Fast-path for intentional Shard CR deletion (`ShardGroupShardForceFail`)](#fast-path-for-intentional-shard-cr-deletion-shardgroupshardforcefail)
      - [EC volume creation - end-to-end round-trip](#ec-volume-creation---end-to-end-round-trip)
      - [EC volume detach - end-to-end round-trip](#ec-volume-detach---end-to-end-round-trip)
      - [EC volume re-attach to a previously-bound node - end-to-end round-trip](#ec-volume-re-attach-to-a-previously-bound-node---end-to-end-round-trip)
      - [EC volume deletion - end-to-end round-trip](#ec-volume-deletion---end-to-end-round-trip)
      - [EC shard failure - end-to-end round-trip](#ec-shard-failure---end-to-end-round-trip)
      - [EC volume expansion - end-to-end round-trip](#ec-volume-expansion---end-to-end-round-trip)
      - [EC engine restart and FrontendSwitchOver](#ec-engine-restart-and-frontendswitchover)
      - [EC snapshot operations - end-to-end round-trip](#ec-snapshot-operations---end-to-end-round-trip)
      - [Auto-salvage for EC volumes](#auto-salvage-for-ec-volumes)
      - [ShardGroup process restart - end-to-end round-trip](#shardgroup-process-restart---end-to-end-round-trip)
      - [Shard eviction and node drain](#shard-eviction-and-node-drain)
      - [Volume expansion (resize, same k+m)](#volume-expansion-resize-same-km)
        - [Three-way comparison of expansion mechanisms](#three-way-comparison-of-expansion-mechanisms)
        - [SPDK patch dependency](#spdk-patch-dependency)
      - [Volume.Status.Robustness mapping](#volumestatusrobustness-mapping)
      - [spec.dataLayout](#specdatalayout)
      - [Volume Deletion](#volume-deletion)
    - [Test plan](#test-plan)
    - [Upgrade strategy](#upgrade-strategy)
  - [API changes](#api-changes)
    - [SPDK JSON-RPC](#spdk-json-rpc)
    - [go-spdk-helper](#go-spdk-helper)
      - [New type definition (`pkg/spdk/types/ec.go`)](#new-type-definition-pkgspdktypesecgo)
      - [New client methods (`pkg/spdk/client/ec.go`)](#new-client-methods-pkgspdkclientecgo)
      - [New CLI commands (`app/cmd/basic/bdev_ec.go`)](#new-cli-commands-appcmdbasicbdev_ecgo)
      - [New `bdev_nvme_reset_controller` wrapper](#new-bdev_nvme_reset_controller-wrapper)
      - [longhorn spdk-engine gRPC (`spdkrpc.SPDKService`)](#longhorn-spdk-engine-grpc-spdkrpcspdkservice)
      - [longhorn-instance-manager changes](#longhorn-instance-manager-changes)
      - [Proxy service impact](#proxy-service-impact)
      - [Disk service impact](#disk-service-impact)
      - [longhorn-manager changes](#longhorn-manager-changes)
  - [Known implementation gaps in initial release](#known-implementation-gaps-in-initial-release)
  - [Current limitation](#current-limitation)
  - [Future improvements](#future-improvements)
```

## Summary

> Longhorn V2 currently replicates entire volumes via RAID1 across nodes - each replica stores a full copy of the data. This provides simple fault tolerance but at Nx storage overhead.

Reviewer: ```suggestion
Longhorn V2 currently replicates entire volumes via RAID1 across nodes - each replica stores a full copy of the data. This provides simple fault tolerance but at (N-1)x storage overhead.
```

or 
```
... This provides simple fault tolerance but requires N× physical storage capacity.
```

> V2 Sharding aims to provide a new storage tier option with erasure coding (EC). A k+m EC configuration splits data into k chunks and m parity chunks using Reed-Solomon coding (ISA-L). Any m simultaneous disk failures can be tolerated while storing only (k+m)/k * the original data, compared to m+1 * for replication. For example:

Reviewer: To avoid confusion about the meaning of `m` in replication.

```suggestion
V2 Sharding aims to provide a new storage tier option with erasure coding (EC). A k+m EC configuration splits data into k data chunks and m parity chunks using Reed-Solomon coding (ISA-L). The system can tolerate up to m simultaneous chunk failures while requiring only (k+m)/k times the original storage capacity. In contrast, full replication requires f+1 complete copies to tolerate f failures.
```

```
A 4+2 EC layer provides 2-failure tolerance using 6 devices at 1.5x storage overhead (150 GB raw for 100 GB usable), compared to 3-way replication with provides 2-failure tolerance using 3 devices overhead (300GB raw for 100GB usable).
```

> This enhancement introduces a new SPDK bdev module (`bdev_ec`) that sits below the lvol/blobstore layer and above the base bdev layer, performing transparent erasure coding on all I/O. The EC bdev's base devices can be remote NVMe-oF attached bdevs on other nodes. This breaks the physical node boundary of the current replication model: instead of each node holding a full replica, chunks are distributed across nodes at the EC layer. Longhorn's control plane sees only a single lvol volume and does not manage chunk placement - the cross-node distribution is entirely transparent at the SPDK layer.

Reviewer: > Longhorn's control plane sees only a single lvol volume and does not manage chunk placement

lvol volume?

Author: Fixed. Thank you!

Fault tolerance comes from Reed-Solomon coding across the distributed base devices.

### Related Issues

- https://github.com/longhorn/longhorn/issues/1061

## Reviewer's overview

This LEP spans SPDK internals, gRPC surfaces across four repos, and Longhorn control-plane orchestration. It is not designed to be read end-to-end. Use this section to orient and pick a reading path.

### Topology in one picture

```
            +----------+                +-------------+               +---------+
 workload --|  Engine  |--- raid1(1) ---| ShardGroup  |--- bdev_ec ---| Shard 0 |
            | (engine  |    NVMe-oF     |   process   |    over k+m   |   ...   |
            |  node)   |                |  (engine    |    NVMe-oF    |  Shard  |
            +----------+                |   node)     |     legs      |  k+m-1  |
                                        | lvstore +   |               +---------+
                                        | head lvol   |               (one per node;

>                                         | live HERE   |                hard anti-affinity)

Reviewer: According to the topology, we only support the hard anti-affinity for shards. In the future, we might support multiple shards on one node.

                                        +-------------+
```

- One Volume maps to one ShardGroup CR (1:1, same name), which maps to k+m Shard CRs.

> - The lvstore and head lvol live on `bdev_ec` **inside the ShardGroup process**, not on the engine. This is what makes detach data-preserving and engine teardown safe.

Reviewer: ~~Could you elaborate more on why it makes detach data-preserving safe?!~~
See the answer in **Six load-bearing design decisions**

> - The engine is structurally a single-replica raid1 aggregator over the ShardGroup endpoint. RAID1 and EC engines are byte-identical inside `SpdkInstanceSpec`; the only discriminator is `data_layout_type` on the gRPC request.

Reviewer: Why do we need the raid1 aggregator? Can we directly expose the EC bdev by the NVMe-of protocol?

Author: This is not an IO path correctness requirement. It's there to keep the engine layout-blind, so RAID1 and EC can use the same engine code path via the upstream interface.

Exposing `bdev_ec` directly and dropping the engine, we would end up splitting things into a second EC-only path. That, I believe, would push frontend management, switchover (live migration), and snapshot/expand orchestration down into the ShardGroup.

IMO, I don't think that extra complexity is worth saving one passthrough layer.

Author: Updated the LEP to state this.

- **Shard CR (new):** represents one base device slot in the EC array. Contains the node, disk, slot index, role (data/parity), health state (normal/failed/replacing), storage IP, and NVMe-oF port. Each EC volume has k+m Shard CRs instead of 2-3 Replica CRs. The Shard CR lifecycle (create lvol -> expose via NVMe-oF -> attach on engine node -> monitor health -> replace on failure -> rebuild) mirrors the Replica CR lifecycle but with EC-specific state tracking.

## Design

**Terminology:**

- **Chunk**: A single unit of data stored on one disk within a stripe. In SPDK, this is referred to as `strip_size` following RAID conventions. This document uses "chunk" instead to avoid confusion with "stripe"
- **Stripe**: One full row across all k+m disks. Each stripe contains k data chunks and m parity chunks.
- **Slot**: A fixed position (0..k+m-1) in the EC array. It is baked into the Reed-Solomon encode matrix and is permanent for the volume's life.
- **RMW (read-modify-write)**: The sub-stripe write path - read the stripe's data chunks, overlay the new bytes, re-encode parity, write back.

> - **WIB (write-intent bitmap)**: Persistent bitmap on the parity disks marking regions that may have a parity-modifying write in flight, so crash recovery only has to scrub those regions.

Reviewer: nit:

```suggestion
- **WIB (write-intent bitmap)**: Persistent bitmap on parity disks marking regions with in-flight or partially applied parity updates. It allows crash recovery to limit scrub/rebuild to only affected regions instead of scanning the full disk.
```

Author: Improving based on suggestion, but scoped to scrub only: the WIB bounds the startup scrub, while the rebuild walks every stripe.

- **Scrub**: Background pass that re-encodes parity for WIB-dirty regions after a crash, making parity consistent with whatever data landed.

> - **ShardGroup process**: The SPDK process that owns `bdev_ec` + lvstore + head lvol for one EC volume - the EC analog of a Replica process in RAID1.

Reviewer: `process` is a bit confusing because user might think it is like a linux process. How about renaming it to `ShardGroup instance`?

#### Threading model

`bdev_ec` is multi-reactor: every I/O path is bracketed by two thread roles.

- **Home thread.** The SPDK thread on which the EC bdev was created. Owns all shared `ec_bdev` state - rebuild, scrub, and resize contexts; persist flags; bit-clear queues - and the long-lived home-owned channels (`wib_chans`, `bitmap_chans`, `rebuild_chans`, `scrub_chans`). RPC handlers, pollers, and persist completion callbacks all run here.
- **Submitter thread.** The SPDK thread on which a `bdev_io` was issued (typically an NVMe-oF poll-group thread). Owns the per-thread `base_chans[]` used to dispatch child reads / writes / unmaps to the base bdevs; child completions land here too, since SPDK delivers each completion back on the channel-owning thread.

Entry points (read / write / UNMAP) route to home before touching shared state; fan-out dispatchers route to the submitter so child I/O is issued on the channel-owning thread; completion routes back to the submitter so the `bdev_io` completes there. Every cross-thread hop uses the same shape: inline call when the caller is already on the target thread, `spdk_thread_send_msg` otherwise.

> **Channel counts.** Each channel set is sized differently: `base_chans[]` — one per thread that issues I/O; `wib_chans[]` — one per parity disk (the WIB is parity-mirrored); `bitmap_chans[]` — one per disk; `rebuild_chans[]` and `scrub_chans[]` — one per live slot during their respective background ops.

Reviewer: Is the flow correct?

<img width="1440" height="1440" alt="Image" src="https://github.com/user-attachments/assets/74084a8a-039f-44ab-b29d-bc85d6fbd776" />

Author: Yes the flow is correct.

#### Mental model: layout and the three maps

Before the mechanisms, the picture they all refer back to. `bdev_ec` is a virtual bdev over k+m base bdevs. Two things anchor everything below: how each disk is laid out, and the three bit-maps the module keeps.

> **On-disk layout (per base disk, uniform across all k+m).** Every disk reserves the same front region for metadata, then user data fills the rest. This is the canonical layout; later sections refer to it rather than redraw it.

Reviewer: What does the **user data** mean? Is it the IO data written by applications or users?

Author: The part of each base disk after the front metadata (bitmap + WIB) that holds this disk's EC chunk per stripe - data chunks on data disks and parity chunks on parity disks. So it is the EC payload, not necessarily application data (parity disks hold parity).

Author: Added to terminology.

| `stripe_unmapped_map` | persistent (front of every disk, raw-replicated, double-buffered) | one stripe | "this stripe is logically zero (unmapped)" - so reads return zeros without touching disk | [How UNMAP works](#how-unmap-works-capacity-reclamation) |
| `wib_region_map` (WIB) | persistent (front of parity disks, double-buffered) | 1024 stripes (a region) | "this region may have a parity-modifying write in flight" - the crash write-hole guard | [WIB on-disk layout](#wib-on-disk-layout) |

The first is a runtime interlock that vanishes on crash (safe - a crash kills the in-flight I/O it guarded). The other two are durable and are what crash recovery and UNMAP correctness rest on.

**The I/O paths, at a glance.** A read returns data from one disk (healthy) or rebuilds it from survivors + parity (degraded), and short-circuits to zeros for an unmapped stripe. A full-stripe write encodes parity from the new data and writes all k+m chunks. A sub-stripe write takes the read-modify-write (RMW) path. UNMAP flips the unmapped bit. The next sections walk each path in turn.

#### Async bdev_ec_create

> `bdev_ec_create` is an asynchronous JSON-RPC: the response (`true`) is sent only after the EC bdev is fully initialized, including the WIB load from disk. This is deliberate - blocking the home thread to read the parity disks synchronously would stall I/O for every other bdev on the system.

Reviewer: `bdev_ec_create` is described as an **asynchronous** JSON-RPC. However, is the RPC response only returned after all initialization steps (including WIB loading) are completed?

What is the expected duration of these initialization steps?

Does the initialization time scale with volume size, stripe size, or the number of data/parity chunks (k+m)?

Is it possible to estimate or model this cost?

Author: > `bdev_ec_create` is described as an **asynchronous** JSON-RPC. However, is the RPC response only returned after all initialization steps (including WIB loading) are completed?
> 

Yes, **asynchronous** just means the call doesn't block on disk I/O. The handler does the in-memory setup, then returns, and the `true` response comes back later via a completion callback once the init finishes.

> What is the expected duration of these initialization steps?
> 
> Does the initialization time scale with volume size, stripe size, or the number of data/parity chunks (k+m)?
> 
> Is it possible to estimate or model this cost?

Most of the cost comes from the unmapped bitmap I/O, not the WIB. The WIB is about 1024x smaller (one bit per 1024 stripes vs one bit per stripe), and the geometry + encode-matrix setup is in memory. 

The bitmap I/O scales with the number of stripes in the volume:
- `num_stripes = volume_size / (k * stripe_size)`
- `bitmap size ≈ num_stripes / 8 bytes`
So the main factor is volume size; increasing `k` or stripe size reduces the number of stripes and therefore the I/O cost.

WIB I/O does not scale the same way  - each op touches a fixed strip-level unit regardless of the volume size. On create, the bitmap I/O:
- read once, and
- on a fresh volume, writes out to all n disks twice.
WIB updates only touch the m parity disks.

I don't have a concrete timing model. But it likely comes down to those bytes plus a few NVMe-oF round-trips to the shards.

Author: Updated the LEP to cover this.

Usable capacity = k / (k+m) = 50% of total raw storage.
```

<a id="slot-index-permanence"></a>**Slot index is permanent.** The position of a disk in `BaseBdevs[]` at `bdev_ec_create` time is its permanent slot assignment for the entire life of the volume. This is not a soft label - it is baked into the Reed-Solomon encode matrix.

ISA-L builds the encode matrix via `gf_gen_rs_matrix` once at `bdev_ec_create`. Each row of the matrix corresponds to one output chunk (data or parity). Each row carries its own set of multipliers, fixed by that row's slot index - so the parity on disk 2 combines the data chunks with one set of Galois-field weights, and the parity on disk 3 combines them with a different set. This mapping is fixed on-disk from the moment the first stripe is written.

Consequences:
- Two disks in the same EC bdev cannot be swapped without re-encoding every parity chunk on disk. Doing so silently corrupts all parity.
- A replacement disk (hot-swap via `bdev_ec_replace_base_bdev`) must occupy the **same slot index** as the disk it replaces. The rebuild writes reconstructed data for that slot to the new disk - slot identity is preserved.

> - The control plane must **persist `slot_index`** in the Shard CR at creation time and pass it faithfully to every subsequent `ShardGroupCreate` and `ShardGroupShardReplace` call. It is not a hint that can be inferred from node layout or disk order at runtime; it is part of the volume's on-disk identity.

Reviewer: In addition to persisting slot_index in the Shard CR, do we also store it on each base bdev?

If all Shard CRs were accidentally deleted, would it still be possible to recover the shard identities from the metadata stored on the base bdevs and reconstruct the Shard CRs?

Author: Not really usable currently:
- `bdev_ec` stores no slot/k/m/role on the base bdev; the "slot" is just the index in the `base_bdevs[]` passed to `bdev_ec_create`.
- The one durable on-disk trace of the slot is the shard's lvol name ( `shard-<volumeName>-<slotIndex>`), but that isn't wired to the reconstruction path.

So if the Shard CRs are gone, there isn’t enough on-disk base bdev metadata to reliably rebuild the volume. For disk self-describing recovery, we probably need to add this metadata later once lvol xattrs are in place.

Author: Cover in future improvements.

#### How address translation works through the stack:

When an application writes to a mounted filesystem, the write passes through several layers, each translating the address before passing it down. Example of how a write to logical byte offset 100 KB traces through the stack (k=2, m=2, chunk_size=32 KB, 4 KB block size).
```
Layer 1: Filesystem (ext4)
    Application writes to file offset -> ext4 maps to logical block number

>     Example: file write -> logical block 25 (= byte offset 100 KB / 4 KB)

Reviewer: ```suggestion
Layer 1: Filesystem (ext4)
    Application writes to a file offset. ext4 translates it to a block on the underlying block device in two steps:
      1. file offset -> file logical block number
         (using the inode's extent tree)
      2. file logical block number -> block device LBA
         (via the extent mapping)
    Example: file write at byte offset 100 KB
             -> file logical block 25 (= 100 KB / 4 KB)
             -> block device LBA 25 (assuming an identity mapping for illustration)
```

Author: Applied, with one fix - the extent tree was on the wrong step.

Layer 2: NVMe-oF initiator (kernel)
    Logical block 25 -> NVMe command send over TCP to spdk_tgt

> 

Reviewer: ```suggestion
Layer 2: Linux block layer / NVMe-oF initiator
    The Linux block layer passes a write request for LBA 25 to the NVMe/TCP initiator, which encapsulates it as an NVMe WRITE command (SLBA = 25) and sends it over TCP to spdk_tgt.
```

Layer 3: NVMe-oF target (spdk_tgt)
    Receives NVMe command for block 25 -> forwards to the lvol bdev

Layer 4: Lvol / Blobstore

>     Lvol 25 -> blobstore looks up which cluster this block belongs to, maps it to a physical offset within the lvol store's address space.

Reviewer: Lvol 25?

Author: Meant lvol block 25, fixed, thank you!

    This is where the EC address translation happens.
    ec0 block -> which stripe? which disk? which offset on that disk?

    chunk_size_in_blocks = chunk_size / block_size
                         = 32 KB / 4KB
                         = 8 blocks per chunk

    stripe blocks = k * chunk_size_in_blocks
                  = 2 * 8
                  = 16 blocks per stripe

    stripe index = 25 / 16
                 = 1 (second stripe, 0-indexed)

    offset_in_stripe = 25 % 16
                     = 9 (9th block within the stripe)

    disk_index = 9 / 8
               = 1 (second data disk, since chunk = 8 blocks)
    
    offset_in_chunk = 9 % 8
                    = 1 (1st block within disk 1's chunk)

    disk_lba = stripe_index * chunk_size_in_block + offset_in_chunk
             = 1 * 8 + 1
             = LBA 9 on disk 1

Layer 6: Base bdev (disk 1 = AIO / NVMe / NVMe-oF)
    LBA 9 -> physical I/O to the underlying storage device
```

The EC layer is the only layer that fans out to multiple disks. All other layers are 1-to-1 address translation. For writes, the EC layer also computes parity and write to the parity disks at the same stripe LBA.

#### How reads work:

> **Threading.** Submitter -> home (shared-state checks: slot-state, unmapped-bitmap read) -> submitter (child reads on `base_chans[]`) -> submitter (completion). See [Threading model](#threading-model).

Reviewer: It is worth adding a description or an example to describe what happens if the bit in the unmapped-bitmap is non-zero.

Author: Good idea. Added, thank you!

> In the normal case (all disks healthy), a read maps directly to on data disk (no parity is involved). If only pariaty disks have failed, reads still go directly to data disk since parity is not needed for reads.

Reviewer: Typo

```suggestion
In the normal case (all disks healthy), a read maps directly to on data disk (no parity is involved). If only parity disks have failed, reads still go directly to data disk since parity is not needed for reads.
```

The current implementation using a fixed separation between data and parity placement. We can introduce distributed parity to balance I/O across disks in the future.

<img width="1440" height="1040" alt="Image" src="https://github.com/user-attachments/assets/e844891d-0783-42a6-a047-6ff25ea2e854" />

Author: Added to future improvement.

> If a data disk has failed, the module reads k chunks from the first k healthy disks (scanning in order, skipping failed ones - this includes data and parity disks). It then reconstructs the missing data chunk using ISA-L Reed-Soloman (RS) decoding (`gf_invert_matrix` to build recovery coefficients, then `ec_encode_data` to apply them).The RS math should guarantee reconstruction succeeds regardless of which specific k disks survives, as long as at least k are healthy (i.e., up to m failures of any combination).

Reviewer: What happens if an I/O error occurs while writing the reconstructed chunk to the replacement disk during rebuild?

Author: A write error aborts the rebuild. The slot stays `REPLACING`, and the partial rebuild is discarded (safe because a `REPLACING` slot is never read until fully rebuilt). The control plane then restarts the rebuild from the beginning.

Author: Added a short explanation for the case.

> When multiple data disk fails simultaneously (e.g., 2 out of 4 in a 4+2 array), the module reconstructs all missing chunks together within the same read request, instead of repeating the full recovery process for each one. The flow is:

Reviewer: What is the expected behavior when 3 disks fail in a 4+2 array (beyond the fault tolerance limit)?

Author: The array goes OFFLINE once failures >m; all reads & writes are rejected at submission time before any I/O is dispatched. The bdev is not deleted - it refuses IO. With only 3 survivors, no stripes can be reconstructed because decoding needs k=4 readables. It recovers only if failed_count drops back to <=m.

    parity 1 = (weight_c * D0) + (weight_d * D1)
    ```
    The weight (a, b, c, d) are fixed number chosen by ISA-L so that any 2 of the 4 chunks (D0, D1, P0, P1) are sufficient to recover the other 2.
    ```
                    D0      D1
    disk 0 (D0):   [ 1       0 ] -> D0 = 1*D0 + 0*D1
    disk 1 (D1):   [ 0       1 ] -> D1 = 0*D0 + 1*D1
    disk 2 (P0):   [ a       b ] -> P0 = a*D0 + b*D1
    disk 3 (P1):   [ c       d ] -> P1 = c*D0 + d*D1
    ```

    We can store this as a table, where each row describes how to produce one disk's chunk from the original data:
    ```
                    D0    D1
    disk 0 (D0):  [ 1     0 ]   -> D0 = 1*D0 + 0*D1  (just store D0 directly)
    disk 1 (D1):  [ 0     1 ]   -> D1 = 0*D0 + 1*D1  (just store D1 directly)
    disk 2 (P0):  [ a     b ]   -> P0 = a*D0 + b*D1  (weighted mix)
    disk 3 (P1):  [ c     d ]   -> P1 = c*D0 + d*D1  (weighted mix)
    ```

    **How recovery works (decoding):**
    Suppose disk 1 (D1) fails. We can read from disk 0, 2 and 3. So the module picks disk 0 and disk 2:
    ```
    disk 0 gives us D0 (directly, it's a data disk)
    disk 2 gives us P0 = a*D0 + b*D1 (in table we stored earlier)
    ```

>     We knoe D0 (we read it) and P0 (we read it). We know that value of a and b because they were generated when the EC array was created and are stored in memory - they are the same value used to compute P0 in the first place. So we can solve for D1:

Reviewer: ```suggestion
    We know D0 (we read it) and P0 (we read it). We know that value of a and b because they were generated when the EC array was created and are stored in memory - they are the same value used to compute P0 in the first place. So we can solve for D1:
```

    ```
         P0 = a*D0 + b*D1
    -> b*D1 = P0 - a*D0
    -> D1   = (P0 - a*D0) / b
    ```

>     That "solving" step is what table inversion (`gf_invert_matrix`) does - it takes the rows for the survivors and computes new recovery wrights that tell us how to combine the survior chunks to get back the missing data. The recovery table is not stored on disk - it is computed on the flow during each degraded read from the original encoding table (which is kept in memory for the lifetime of the EC array). It must be recomputed each time because which disk have failed can change at any moment.

Reviewer: ```suggestion
    That "solving" step is what table inversion (`gf_invert_matrix`) does - it takes the rows for the survivors and computes new recovery weights that tell us how to combine the survivor chunks to get back the missing data. The recovery table is not stored on disk - it is computed on the fly during each degraded read from the original encoding table (which is kept in memory for the lifetime of the EC array). It must be recomputed each time because which disks have failed can change at any moment.
```

    **The multi-failure is the same idea**
    If 2 disk fail (say disk 0 and disk 1 - both data disks gone), we read from disk 2 (P0) and disk 3 (P1)
    ```
    P0 = a*D0 + b*D1
    P1 = c*D0 + d*D1
    ```

>     Two equations, two unknown (D0 and D1). The table inversion solves both simultaneously, producing recovery weight that say "combine P0 and P1 with these new weight to get D0" and "combine P0 and P1 with these other weight to get D1". ISAL applies both recipes to the survivor data in one pass.

Reviewer: ```suggestion
    Two equations, two unknown (D0 and D1). The table inversion solves both simultaneously, producing recovery weights that say "combine P0 and P1 with these new weight to get D0" and "combine P0 and P1 with these other weights to get D1". ISAL applies both recipes to the survivor data in one pass.
```

>     The RS algorithm guarantees this system of equations is wasy solvable as long as we have at least k survivors - no matter which specific disk failed. This is why up to m failures can be tolerated.

Reviewer: ```suggestion
    The RS algorithm guarantees this system of equations is easy solvable as long as we have at least k survivors - no matter which specific disk failed. This is why up to m failures can be tolerated.
```

> 1. **Extract recovery recipes**: From the inverted table, extract one row per failed disk. Each row is a "receipt" - a set of weights that says how to combine the k survivor chunks to reproduce that specific missing chunk. If 2 disk failed, we extract 2 recipes.

Reviewer: ```suggestion
1. **Extract recovery recipes**: From the inverted table, extract one row per failed disk. Each row is a "receipt" - a set of weights that says how to combine the k survivor chunks to reproduce that specific missing chunk. If 2 disks failed, we extract 2 recipes.
```

> 1. **Recovery all at once**: Feed al the receipt and the k survivor chunks into ISA-L (`encode_data` - the same function used for encoding, since the math is identical; the difference is which coefficient are used) in one call. ISA-L applies each recipe to the survivor data and produces all missing chunks simultaneously.

Reviewer: ```suggestion
1. **Recovery all at once**: Feed all the receipt and the k survivor chunks into ISA-L (`encode_data` - the same function used for encoding, since the math is identical; the difference is which coefficients are used) in one call. ISA-L applies each recipe to the survivor data and produces all missing chunks simultaneously.
```

#### How write works

**Threading.** Submitter -> home (stripe-busy claim, active-scrub guard, WIB persist setup) -> submitter (k+m child writes on `base_chans[]`) -> submitter (completion). The RMW path follows the same pattern; the read phase and the parity-write phase both fan out on the submitter. See [Threading model](#threading-model).

> A full-stripe write (the entire stripe is being written at once) participates in the same WIB protocol as RMW. (A write that lands on a stripe currently marked unmapped is also handled as a full-stripe write, but it skips the WIB and instead clears the stripe's unmapped bit after the data lands - see [Writing into an unmapped stripe](#writing-into-an-unmapped-stripe-clearing-a-bit) below.)

Reviewer: How about

```suggestion
A full-stripe write (where the entire stripe is written at once) follows the same WIB protocol as RMW. The WIB ensures that, after a crash, recovery knows which stripes may have had an in-flight parity-modifying write and therefore need to be scrubbed.

If the target stripe is currently marked unmapped, however, the WIB is skipped. Because the stripe is logically all zeros and the write completely replaces its contents, a crash cannot leave the stripe in a state where old and new data are mixed. The data is written directly, and the stripe's unmapped bit is cleared once the write completes. Skipping the WIB in this case also avoids unnecessary metadata updates and reduces write-path overhead. See [Writing into an unmapped stripe](#writing-into-an-unmapped-stripe-clearing-a-bit).
```

Reviewer: I just noticed you already mentioned it below, but it would be clearer to introduce it earlier.

Author: Good idea. Moved the unmapping case into its own sentence so it shows up earlier. I made some tweaks from your suggestion to keep things accurate.

1. Claim the stripe-busy bit (`ec_stripe_set_dirty`).

> 1. Active-scrub guard (defer via NOMEM if a scrub is in flight on this stripe's region at-or-ahead of the request).

Reviewer: Using NOMEM is a bit weird. Why not use EAGAIN or EBUSY?

Reviewer: Another question, will the error code be returned back to the application?

Author: > Using NOMEM is a bit weird. Why not use EAGAIN or EBUSY?

I should make it clear that these are `spdk_bdev_io_status`, not errno.

`EAGAIN` and `EBUSY` are not `bdev_io` [completion statuses](https://github.com/longhorn/spdk/blob/440f10735b5c2c2a7c7285889b66434995932181/include/spdk/bdev_module.h#L371), and `NOMEM` is the only one triggers auto-requeues path:
- https://github.com/longhorn/spdk/blob/440f10735b5c2c2a7c7285889b66434995932181/lib/bdev/bdev.c#L1719-L1721
- https://github.com/longhorn/spdk/blob/440f10735b5c2c2a7c7285889b66434995932181/lib/bdev/bdev.c#L1650-L1657

We do use `-EAGAIN`, but intentionally by [dispatching to  SPDK_BDEV_IO_STATUS_NOMEM](https://github.com/c3y1huang/longhorn-spdk/blob/f2aba8467e94bde9bcaa1704fa776fe7dc6f5323/module/bdev/ec/bdev_ec.c#L2077-L2083).

I've tweaked the line to make it less confusing.

Author: > Another question, will the error code be returned back to the application?

No, the `SPDK_BDEV_IO_STATUS_NOMEM` returned for a deferred write isn't surfaced back to the application. It stays inside the SPDK bdev layer - requeues, and retries.

1. Mark the WIB region dirty in memory, increment `wib_region_inflight[region]`, refresh `wib_region_dirty_ts[region]`.
1. Persist the WIB bit to disk if needed (same three-branch decision).
1. Read all k data chunks for the stripe from healthy disks.
1. If any data disk has failed, reconstruct the missing chunks from the surviving k disks using RS decoding (`gf_invert_matrix` + `ec_encode_data`).
1. Copy the caller's write data into the data buffer at the offset corresponding to the write's position within the stripe. For example, if the stripe covers logical blocks 0-63 and the write targets block 20-25, the new data is copied into the buffer at the position for block 20-25, leaving the rest of the stripe's data unchanged.
1. Re-encode all m parity chunks from the modified data (`ec_encode_data`).
1. Write the modified data chunk and all m parity chunks back to disk. The other k-1 data chunks are not written back - they were read for parity encoding but their on-disk content is unchanged, so writing them would be wasted I/O. This reduces write amplification from k+m writes per RMW to 1+m writes.
1. Decrement `wib_region_inflight[region]`. Clear the stripe-dirty bit.

> 

Reviewer: ```mermaid
flowchart TD
    S1["1. Claim stripe-dirty bit\nrequeue if already dirty"]
    S2["2. Mark WIB region dirty\nincrement inflight, refresh dirty_ts"]
    S3["3. Persist WIB bit to disk\nalready on disk / pending / clean"]
    S4["4. Read all k data chunks"]
    D1{"data disk failed?"}
    S5["5. Reconstruct missing chunks\ngf_invert_matrix + ec_encode_data"]
    S6["6. Copy new data into buffer"]
    S7["7. Re-encode m parity chunks\nec_encode_data"]
    S8["8. Write 1 data + m parity\n1+m writes, not k+m"]
    S9["9. Release stripe-dirty bit\ndecrement inflight, complete bdev_io"]

    S1 --> S2 --> S3 --> S4 --> D1
    D1 -- yes --> S5 --> S6
    D1 -- no --> S6
    S6 --> S7 --> S8 --> S9
```

> 

Reviewer: ```mermaid
flowchart TD
    S1["**1. Claim stripe-dirty bit**\nec_stripe_set_dirty — if already dirty, requeue"]
    S2["**2. Mark WIB region dirty**\nincrement wib_region_inflight, refresh wib_region_dirty_ts"]
    S3["**3. Persist WIB bit to disk**\nalready on disk → fanout\npersist pending → defer with wib_repersist_needed\nclean → start persist, resume on completion"]
    S4["**4. Read all k data chunks**\nfrom healthy disks in parallel"]
    D1{"any data disk failed?"}
    S5["**5. Reconstruct missing chunks**\ngf_invert_matrix + ec_encode_data"]
    S6["**6. Copy new data into buffer**\noverwrite target offset, rest unchanged"]
    S7["**7. Re-encode parity**\nec_encode_data over all k data chunks → m parity chunks"]
    S8["**8. Write 1 data + m parity chunks**\nskip unchanged k−1 data chunks\nreduces write amplification from k+m to 1+m"]
    S9["**9. Release stripe-dirty bit**\ndecrement wib_region_inflight, complete bdev_io"]

    S1 --> S2 --> S3 --> S4 --> D1
    D1 -- yes --> S5 --> S6
    D1 -- no --> S6
    S6 --> S7 --> S8 --> S9
```

+-------------------------------------------------------+

array[1] holds bits for stripes 64-127:
+-------------------------------------------------------+
| bit0  bit1  bit2  bit3  ............... bit62  bit63  |
| str64 str65 str66 str67 ............... str126 str127 |
+-------------------------------------------------------+

array[2] holds bits for stripes 128-191...
...
```

To find which bit represents a given stripe, two steps:
- Which array element? Divide the stripe number by 64, drop the remainder. This gives us which group of 64 the stripe falls in.
- Which bit within that element? Subtract the group's starting stripe number. This gives us the position within the group.

Example:
```
Stripe 100:
    100 // 64 = 1    -> array[1] (covers stripe 64-127)
    100 - 64  = 36   -> bit 36 inside array[1]

Stripe 230:
    230 // 64 = 3    -> array[3] (covers stripe 192-255)
    230 - 192 = 38   -> bit 38 inside array[3]

Stripe 5:
    5 // 64 = 0      -> array[0] (covers stripe 0-63)
    5 - 0   = 5      -> bit 5 inside array[0]
```
The bit operations:
```
Set dirty:      array[1] |=  (1 << 36)
Clear dirty:    array[1] &= ~(1 << 36)
Check dirty:    array[1] &   (1 << 36)
```

> The bitmap is allocated when the EC array is created, sized to cover all `num_stripes` *user* stripes. (The front-of-disk bitmap-region strips are not tracked here - they are raw-replicated, not EC-encoded, so there is no bitmap-region rebuild walk.) A single global `bitmap_persist_in_flight` flag serializes any in-flight bitmap persist. The stripe indices in this section are user-stripe indices; the `+data_offset_stripes` translation to a physical base LBA happens only at `ec_stripe_base_lba()`, not in `stripe_dirty_map`. All bits start at 0 (clean). No locking - mutations are home-only and the submitter-side reader follows the access discipline in [Threading model](#threading-model).

Reviewer: ```suggestion
The bitmap is allocated when the EC array is created, sized to cover all `num_stripes` *user* stripes. (The front-of-disk bitmap-region stripes are not tracked here - they are raw-replicated, not EC-encoded, so there is no bitmap-region rebuild walk.) A single global `bitmap_persist_in_flight` flag serializes any in-flight bitmap persist. The stripe indices in this section are user-stripe indices; the `+data_offset_stripes` translation to a physical base LBA happens only at `ec_stripe_base_lba()`, not in `stripe_dirty_map`. All bits start at 0 (clean). No locking - mutations are home-only and the submitter-side reader follows the access discipline in [Threading model](#threading-model).
```

> Before starting RMW, the module checks the stripe bit - if its dirty (another RMW is in progress), then requeue the write. If is clean, the module sets it dirty to claim exclusive access to the stripe.

Reviewer: ```suggestion
Before starting RMW, the module checks the stripe bit - if it is dirty (another RMW is in progress), then requeue the write. If it is clean, the module sets it dirty to claim exclusive access to the stripe.
```

Reviewer: > Before starting RMW, the module checks the stripe bit - if its dirty (another RMW is in progress), then requeue the write. 

Could you elaborate on what "requeue the write" means in practice?

Author: In practice, this is SPDK's built-in retry path. When a stripe's dirty bit is already set, the module completes the request with -ENOMEM instead of running the write. The bdev layer then holds onto the I/O and resubmits it once another I/O on the same channel completes. So the caller sees a bit more latency on that write, not a failure.

> Before starting RMW, the module checks the stripe bit - if its dirty (another RMW is in progress), then requeue the write. If is clean, the module sets it dirty to claim exclusive access to the stripe.

Reviewer: ```suggestion
Before starting RMW, the module checks the stripe bit - if it is dirty (another RMW is in progress), then requeue the write. If is clean, the module sets it dirty to claim exclusive access to the stripe.
```

> Before starting RMW, the module checks the stripe bit - if its dirty (another RMW is in progress), then requeue the write. If is clean, the module sets it dirty to claim exclusive access to the stripe.

Reviewer: > if it is dirty (another RMW is in progress), then requeue the write.

Is it possible for a write to be requeued indefinitely? Do we need a maximum retry count?

Author: > Is it possible for a write to be requeued indefinitely?

No, I don't think so. A stripe is held only for one RMW and is freed as soon as it finishes. The only theoretical edge case would be a workload hammering the same stripe. Even then, write should still make forward progress, possibly with higher latency, but won't be stuck. This is the same slowdown-and-retry backpressure that SDPK uses elsewhere.

> Do we need a maximum retry count?

I wouldn't. A retry cap means eventually giving up and failing a write, even when nothing is broken; the stripe is just busy. The `rmw_deferred_inflight` gives visibility. We can see if this becomes a real issue in practice.

#### Crash safety of the dirty bitmap

The dirty bitmap is in-memory only - it does not survive a node crash or reboot. On restart, all bits start at zero (clean). This is safe for bitmap's primary purpose (preventing concurrent RMW on the same stripe) because the crash kills all in flight I/O, so there are no conflicting operations when the node restarts.

> However, there is a separate crash-saftly concerns: If the node crashes mid-RMW - after writing some chunks of a stripe to a disk but not all - that stripe has inconsistent data and parity on disk. For example, the data chunk may bave een updated but the parity chunk was not uet written. After restart, the bitmap is gone so the module doesn't know which stripes were mid-write. A subsequent disk failure could cause silent data corruption because the stale parity doesn't match the updated data.

Reviewer: ```suggestion
However, there is a separate crash-safety concerns: If the node crashes mid-RMW - after writing some chunks of a stripe to a disk but not all - that stripe has inconsistent data and parity on disk. For example, the data chunk may have been updated but the parity chunk was not yet written. After restart, the bitmap is gone so the module doesn't know which stripes were mid-write. A subsequent disk failure could cause silent data corruption because the stale parity doesn't match the updated data.
```

> The following approach is implemented:

Reviewer: ```suggestion
To address this, the module uses a persistent WIB:
```

> - Write-intent bitmap (WIB, persistent on disk) - a persistent bitmap stores on the parity disk that marks which regions have writes in progress. On crash recovery, only the marked regions need scrubbing, not the entire volume. this is the same pattern used by Linux md-raid ( see [kernel RAID document](https://docs.kernel.org/admin-guide/md.html) and [md-raid write intent bitmap wiki](https://archive.kernel.org/oldwiki/raid.wiki.kernel.org/index.php/Write-intent_bitmap.html)). It combines fast startup with simple scrubbing. The small write overhead during normal operation (the on-disk bitmap must be updated before each RMW) is acceptable because it only applies to the first RWM into a cold region - subsequent RMW's to the same region skip the persist (the bit is already set).

Reviewer: ```suggestion
- Write-intent bitmap (WIB, persistent on disk) — a persistent bitmap stored on the parity disk that marks which regions have writes in progress. On crash recovery, only the marked regions need scrubbing, not the entire volume. This is the same pattern used by Linux md-raid (see [kernel RAID document](https://docs.kernel.org/admin-guide/md.html) and [md-raid write intent bitmap wiki](https://archive.kernel.org/oldwiki/raid.wiki.kernel.org/index.php/Write-intent_bitmap.html)). It combines fast startup with simple scrubbing. The small write overhead during normal operation (the on-disk bitmap must be updated before each RMW) is acceptable because it only applies to the first RMW into a cold region — subsequent RMWs to the same region skip the persist (the bit is already set).
```

#### WIB on-disk layout

> The WIB is stored as two double-buffered copies at fixed front offsets, immediately after the in-band unmapped bitmap reservation and before user data. Per base disk:

Reviewer: The benefit of double-buffering is explained later, but worth calling out upfront before diving into the layout details.
```suggestion
The WIB is stored as two double-buffered copies at fixed front offsets, immediately after the in-band unmapped bitmap reservation and before user data. Double-buffering ensures that a crash during a WIB persist always leaves one valid copy intact. Per base disk:
```

Author: Good call, I made a small tweak to remove the redundant wording.

                                                (bitmap_reservation_stripes + copy) * strip_size
```

In the diagram, `0` is LBA 0 at the front of the disk and `total` is `first->blockcnt`, the base disk's full block count; the three regions partition that whole `0 .. total` range.

The WIB reservation is `2 * strip_size` per disk. It is reserved uniformly on **all k+m disks** (data and parity) so the geometry math is single-rule; only parity disks actually write WIB bytes there. The position is a pure function of `strip_size` and the bitmap reservation (both fixed-max at create), so the WIB never moves on resize.

`ec_wib_lba` is declared `static inline` in `bdev_ec_internal.h` because the result is pure arithmetic with no I/O or descriptor lookups, making it cheap to inline at every call site and trivial to exercise in unit tests without pulling in the rest of `bdev_ec_wib.c`.

On-disk header per copy:

```
[ magic       u64 ]   - "EC WIB0" (0x45432057494230)
[ version     u32 ]   - 1 (EC_WIB_VERSION; on-disk format version, bumped in
                          lock-step with EC_BITMAP_VERSION on any layout change;
                          a mismatch is rejected on load)

> [ generation  u32 ]   - monotonically increasing; highest wins on startup

Reviewer: Just out of curiosity, why no use u64 for generation?

Author: No strong reason for u32; u64 is the right call. Good catch, I will update it. Thank you.

[ num_regions u32 ]   - ceil(num_stripes / 1024)

> [ _pad        u32 ]   - reserved; aligns region_bits[] to an 8-byte (uint64_t) boundary

Reviewer: > [ _pad        u32 ]   - reserved; aligns region_bits[] to an 8-byte (uint64_t) boundary

Suggest using `reserved`

Author: Updated.

[ region_bits[]   ]   - ceil(num_regions / 64) uint64_t words, one bit per region

> [ crc32c      u32 ]   - covers everything above

Reviewer: Suggest renaming to `csum`. The field name does not need to encode the algorithm. If it ever changes, the name becomes misleading without a format-breaking rename. 

```suggestion
[ csum   u32 ]   - crc32c checksum; covers everything above
```

Author: Good call. One refinement: SPDK on-disk metadata uses `crc`, e.g.:
- https://github.com/longhorn/spdk/blob/440f10735b5c2c2a7c7285889b66434995932181/lib/blob/blobstore.c#L1734
- https://github.com/longhorn/spdk/blob/440f10735b5c2c2a7c7285889b66434995932181/module/bdev/raid/bdev_raid.h#L607

So I would prefer to stick with that to stay consistent with SPDK's house style.

Author: Updated.

```

> Writes alternate between copy 0 and copy 1 (double-buffering). The active copy is flipped only after all m parity disk writes complete successfully. This ensures the previous copy remains valid if a crash interrupts the persist. On startup, `ec_wib_load` reads both copies from all m parity disks, validates each (magic, version, CRC), picks the highest-generation valid copy from each disk, and OR-merges across all disks - a region is dirty if any parity disk's copy has the bit set.

Reviewer: If both copies are corrupt, what can we do?

If only one copy is corrupt, do we immediately repair it by copying the healthy copy over?

Author: Good questions.

> If only one copy is corrupt, do we immediately repair it by copying the healthy copy over?

We don't rewrite the good copy over it at load - load only reads. We take the valid copy with the highest generation and leave the bad one alone. Each new write goes to the other of the two slots, so the next WIB write lands on the bad slot and overwrites. And if the whole parity disk's WIB is unreadable, every parity disk holds the same WIB, so we will read it from the good one.

> If both copies are corrupt, what can we do?

There is currently a gap here - I've missed it. ATM, load assumes clean and skips scrub. Meaning if a volume loses its whole WIB, it will still come up with no scrub, potentially leaving parity inconsistent.

---

I will address and split via the salvage flag:
- Fresh volume: assume clean, nothing to scrub.
- Existing volume (salvage): mark all regions dirty and do a full scrub, re-encode parity from the data disks.

Author: Updated.

                           +- bytes left in one strip after the on-disk header (24 B) + CRC32C (4 B) = 28 B
```

In words: count the bytes available for the bitmap inside one strip, multiply by 8 to bet bits, by 1024 to get the stripes those bits address, and by the per-stripe user-data size to get total addressable capacity. `ec_compute_geometry` enforces this bound and refuses to register the bdev if exceeded.

| Strip size          | k=2       | k=4         | k=8         | k=16        |
| ------------------- | --------- | ----------- | ----------- | ----------- |
| 4 KB                | 0.2 TiB   | 0.5 TiB     | 1.0 TiB     | 2.0 TiB     |
| 16 KB               | 4 TiB     | 8 TiB       | 16 TiB      | 32 TiB      |
| **64 KB (default)** | **64 TiB**| **128 TiB** | **256 TiB** | **512 TiB** |
| 256 KB              | 1 PiB     | 2 PiB       | 4 PiB       | 8 PiB       |
| 1 MB                | 16 PiB    | 32 PiB      | 64 PiB      | 128 PiB     |

The two scaling axes compose: doubling k doubles the max volume size, and doubling strip_size *quadruples* it - both the bytes-per-bit and the bits-that-fit-in-one-strip grow linearly with strip size.

> For the default Longhorn EC configuration (4+2 with 64 KB strips), the limit is ~128 TiB per volume. Larger volumes need either a larger strip size at create time. The limit cannot be raised after creation, because changing `strip_size_kb` would invalidate the parity layout.

Reviewer: Is it possible for the WIB to span two strips, and would that allow the maximum volume size to increase to 256 TB with a 64 KB strip size and k=4? The most straightforward way to increase this limit is to enlarge the strip size, but this introduces higher write amplification due to increased RMW overhead.

Author: Yes, it's possible, and your number is right. That said, I prefer to keep it out of the initial release and treat it as a future improvement unless we know there is a specific ceiling to meet. For volume >128 TiB, increasing k is probably the better lever for now. WDYT?

**Practical range: 4 KB to 1 MB.** The max-volume-size table above covers this range. Outside it, thing break down:
- *Below 4 KB:* RMW overhead dominates. Small writes pay full read-modify-write cost without getting the throughput benefit of larger I/O.
- *Above 1 MB:* sub-stripe write amplification gets painful (a 4 KB random write at k=4 reads ~4 MB to recompute parity), the WIB scrub blast radius exceeds 4 GiB per dirty region, and DMA buffer pressure rises sharply. The code does not reject these values, but operational properties have not been validated there.

**Choosing strip size by workload:**
| Strip size                 | When it fits                                                                  |
| -------------------------- | ----------------------------------------------------------------------------- |
| Smaller (4-16 KB)          | Small random writes (databases, VMs with many small files). Smaller RMW read penalty. More stripes per volume -> smaller max addressable size. |
| **Default (64 KB)**        | General-purpose or mixed workloads. Reasonable across the board.              |

> | Larger (256 KB - 1 MB)     | Large sequential writes (media, backup targets, archives). Better throughput and larger max volume size, at the cost of higher RMW penalty for sub-stripe writes. |

Reviewer: We can run a benchmark across different strip sizes to evaluate performance impact. This should focus on RMW-heavy workloads (e.g., random small writes and mixed R/W) and compare IOPS, latency (especially tail latency), throughput, and CPU usage. A full-stripe write test can be used as a baseline to separate EC throughput from RMW overhead.

| **4+m (default)** | 4  | **256 MiB**   | **256 MiB**        | **~128 TiB**    |
| 6+m               | 6  | 384 MiB       | 384 MiB            | ~192 TiB        |
| 8+m               | 8  | 512 MiB       | 512 MiB            | ~256 TiB        |
| 16+m              | 16 | 1 GiB         | 1 GiB              | ~512 TiB        |

###### How to choose k, m, and strip_size

Pick parameters in this order, since each depends only on the ones above
it:

1. **m - durability.** How many simultaneous disk failures must the
   volume tolerate? Pick m to match. Independent of WIB sizing.
2. **k - efficiency vs. fan-out.** Higher k means better storage
   efficiency (`k / (k+m)` of raw capacity is usable) but more disks to
   touch on every write. Higher k also requires more schedulable nodes
   for anti-affinity.
3. **strip_size - workload shape.** Large sequential writes want bigger
   strips. Small random writes want smaller strips. Mixed or unknown
   workloads should stay on the 64 KB default.
4. **Verify max volume size.** Cross-check the chosen (k, strip_size)
   against the capacity table. If the volume needs more, doubling

>    strip_size quadruples the ceiling.

Reviewer: We can create a calculator to help users understand the tradeoff between the factors

<img width="1440" height="926" alt="Image" src="https://github.com/user-attachments/assets/b70174eb-97e7-4ddf-b641-c96bb22ce2e8" />

| 100 TiB - 1 PiB   | 4+2, 256 KB    | 1.5x     | Same durability, wider strip raises the ceiling to ~2 PiB. RMW read amplification grows 4x but stays manageable. Scrub blast radius grows to 1 GiB (still seconds on local NVMe).         |
| 1 PiB - 5 PiB     | 8+2, 256 KB    | 1.25x    | Storage efficiency matters at PiB scale. Wider fan-out (10 shards) assumes a mostly cold/archival workload where the RMW path is rarely exercised. Needs 10+ schedulable nodes.           |
| 5 PiB+            | 16+4, 1 MB     | 1.25x    | Enterprise cold storage. Minutes-long scrub windows and wide rebuild fan-out; only appropriate for mostly-sequential workloads (bulk backup, media archive).                              |

Two caveats when reading this table. Every row is a *creation-time*
decision - `strip_size_kb` and the (k, m) shape are immutable for the
volume's life, so size for the volume's lifetime capacity, not just
provisioning-day capacity. And the overhead column is raw-to-usable ratio
only; it does not include filesystem overhead, snapshot reserves, or

> replica placement at the Longhorn layer.

Reviewer: ```suggestion
Two caveats when reading this table. Every row is a **creation-time** decision - strip_size_kb and the (k, m) shape are immutable for the volume's life, so size them for the volume's lifetime capacity, not just provisioning-day capacity. And the overhead column is the raw-to-usable ratio only; it does not include filesystem overhead, snapshot reserves, or replica placement at the Longhorn layer.
```

What is **provisioning-day capacity**?

Author: > What is provisioning-day capacity?

The size of the volume at creation. Reworded.

#### WIB persist lifecycle

**Threading.** WIB persist (the `wib_persist_in_flight` flag, the `wib_deferred_writes` queue, the idle poller) runs on home over the home-owned `wib_chans[]`. When the persist callback fires, the triggering RMW re-enters its data fan-out on the submitter. See [Threading model](#threading-model).

> 1. Region goes cold -> dirty: The first RMW into a clean region sets the in-memory dirty bit and persists it to disk before any data/parity writes proceed. The RMW's data writes are deferred until the persist confirm the bit on disk (via `ec_rmw_wib_set_cb`).

Reviewer: Is the WIB persisted to parity disks only, or do data disks also need to be updated before the data/parity writes can proceed?

```suggestion
1. Region goes cold -> dirty: The first RMW into a clean region sets the in-memory dirty bit and persists it to the parity disks before any data/parity writes proceed. The RMW's data writes are deferred until the persist confirms the bit on disk (via ec_rmw_wib_set_cb).
```

Author: Fixed, thank you.

> 1. Persist coaliescing: If a persist is already in flight (`wib_persist_in_flight`), the new dirty bit is recorded in memory and `wib_repersist_needed` is set. Data writes for the new RMW are queued on `wib_deferred_writes` (a TAILQ). When the in-flight persist completes, `ec_wib_persists_write_cb` starts a follow-up persist to capture the new bit, then drains the derred queue once the follow-up completes. This guarantees that every dirty bit reaches disk before its corresponding data wries, evnen under concurrent RMW to different cold regions.

Reviewer: Does this mean that all WIB persists and data/parity writes are strictly serialized?

Author: > Does this mean that all WIB persists and data/parity writes are strictly serialized?

Not strictly.

WIB persistence is serialized only during state transitions. A persist happens when a region first flips from clean -> dirty (and again when cleared). It does not happen on every write. Once the WIB bit is set on disk, subsequent writes to that region skip WIB persistence.

Data and parity writes are not serialized. They proceed in parallel across different stripes. Writes only serialize when they hit the same stripe.

> 1. Region goes idle -> clear: A background poller (`ec_wib_idle_poller_cb`, every 100ms) scans regions for those with the dirty bit set, in-flight could of zero, and idle for more than 500 ms (`EC_WIB_IDLE_MS = 500`). It clears the in-memory bit and fires a persit to write the cleared state to disk.

Reviewer: I haven't checked the implementation yet, but how does `ec_wib_idle_poller_cb` ensure that no new dirty bit is written between the idle check and the persist completing? If a new RMW marks the region dirty in that window, could the subsequent persist accidentally clear a bit that should have stayed set?

Reviewer: Could you briefly explain the tradeoff between longer and shorter values for the background poller interval (currently 100ms) and EC_WIB_IDLE_MS (currently 500ms)?

Author: > I haven't checked the implementation yet, but how does `ec_wib_idle_poller_cb` ensure that no new dirty bit is written between the idle check and the persist completing? If a new RMW marks the region dirty in that window, could the subsequent persist accidentally clear a bit that should have stayed set?

Idle clear isn't a separate path - it uses the same `ec_wib_persist` as the RMW. And the in-memory bit changes all run on the home thread, so clearing the bit and a new write re-marking it dirty don't race.

Author: > Could you briefly explain the tradeoff between longer and shorter values for the background poller interval (currently 100ms) and EC_WIB_IDLE_MS (currently 500ms)?

Both values are basically trading off the same things: normal write overhead vs crash recovery cost.

EC_WIB_IDLE_MS (500ms) - the idle threshold before a region's dirty bit is cleared (the region must have no in-flight writes and have been dirty at least this long).
- Shorter: region clear faster -> fewer dirty regions -> faster crash recovery, but bursty workloads might keep flipping the bit on/off, adds a bit more metadata writes.
- Longer: fewer re-marking writes during short pauses, but more dirty regions accumulate -> more work to scrub after a crash.

Poll interval (100ms) - how often the background scan looks for idle regions to clear.
- Shorter: clears bits sooner but scans more often -> more CPU costs.
- Longer: cheaper, but clearing gets delayed.

I think the values are reasonable based on some local runs, and 100 ms is also the value SPDK puts for its other background pollers. That said, I am not sure how well they hold up in the real world. We can tune this if needed.

#### Startup scrub:

**Threading.** The scrub poller runs on home; per-stripe reads fan out on the submitter; completion re-anchors on home to advance scrub state. The follow-up WIB persist that clears region bits also runs on home. See [Threading model](#threading-model).

> On startup, if `ec_wib_load` finds any dirty regions, `ec_bdev_start_scrub` drives a background scrub. For each dirty region, for each stipe in that region: read k data chunks from readable disks, re-encode all m parity chunks, and write parity back. After all tripes in a region are scrubbed, the region bit is cleared. After all dirty regions are scrubbed, the cleared bitmap is persisted.

Reviewer: Is the region bit cleared and persisted immediately after each region is scrubbed? If yes
```suggestion
On startup, if `ec_wib_load` finds any dirty regions, `ec_bdev_start_scrub` drives a background scrub. For each dirty region, for each stipe in that region: read k data chunks from readable disks, re-encode all m parity chunks, and write parity back. After all tripes in a region are scrubbed, the region bit is cleared and persisted. After all dirty regions are scrubbed, the cleared bitmap is persisted.
```

Author: Good question - it's not persisted per region. When a region's stripes are all scrubbed, the bit is cleared in memory; the on-disk write happens at once at the end via `ec_scrub_finish`.

Author: Updated to make it clearer.

The bdev is registered and avoiable for read immediately. RMW writes to a strip whose region has not yet been scrubbed return `EAGAIN` (NOMEM -> requeue by SPDK) to prevent the scrubber and foreground RMW from racing on the same parity data.

> The scrub requires all k data to be NORMAL (so it can read correct data). If any data disk is not NORMAL at startup - FAILED or REPLACING (volume reopened mid-rebuild) - the scrub is deferred until the disk is replaced and rebuilt - the dirty region bits remain set and scrub runs automatically after `ec_rebuild_finish` completes. This is by design and not fixable without adding a persistent scrub journal - the scrub needs correct data to re-encode parity, and with a failed data disk the only way to get the data is through RS reconstruction from surviving disks, which itself relies on parity being consistent (the very thing the scrub is trying to fix). The circular dependency means deferral is the only safe option.

Reviewer: How about making it more clear to readers?

```suggestion
Scrub requires all k data disks to be NORMAL so it can read correct data directly. If any data disk is FAILED or REPLACING at startup (for example, a volume reopened mid-rebuild), the scrub is deferred until the disk is replaced and the rebuild completes. The dirty region bits remain set during this period, and the scrub runs automatically once `ec_rebuild_finish` completes.

This deferral is by design. One might ask: if a data disk has failed, why not reconstruct the missing data via RS decoding from the surviving data and parity disks and proceed with the scrub?

The answer is that after a crash mid-RMW, parity on disk may be partially updated or inconsistent due to in-progress writes. RS reconstruction depends on parity being trustworthy; using potentially stale parity to reconstruct data, then re-encoding parity from that reconstructed data, risks propagating or amplifying existing inconsistency instead of correcting it.

This circular dependency means deferral is the only safe option without introducing a persistent scrub journal to track which stripes have already been verified as consistent.
```

Author: Good call - updated with a couple of tweaks.

> #### How UNMAP works (capacity reclamation)

Reviewer: TODO: Skip this UNMAP. I'm focusing on read/write I/O paths, startup, rebuild, and scrub, rather than less common paths.

- DEGRADED (`0 < failed_count <= m`): One or more disks have failed, but within the RS tolerance limit. Reads reconstruct missing data from surviving disks. Writes skip failed disks (parity is still computed from available data).
- OFFLINE (`failed_count > m`): Too many failures. All new I/O is rejected with an error. The bdev still exists but is unusable until enough disks are replaced and rebuilt.

```
                disk fails           disk fails
             (failed_count++)     (failed_count++)
                    |                    |
                    v                    v
+--------+     +----------+     +------------------+
| ONLINE |---->| DEGRADED |---->|     OFFLINE      |
|   (0)  |     |  (1..m)  |     | (>m, reject I/O) |
+--------+     +----------+     +------------------+
     ^               ^                    ^
     |               |                    |
     |       replace + rebuild    replace + rebuild
     |       (failed_count--)     (failed_count--)
     |               |                    |
     +---------------+<-------------------+
```

> #### How disk failure is handled:

Reviewer: ```suggestion
#### How disk failure is handled
```

- **Computation.** `refreshECSnapshotMapNoLock` populates per-lvol `ActualSize = NumAllocatedClusters * sg.ClusterSize`, then sets `sg.ActualSize = newHead.ActualSize + sum(snapshot ActualSize)`.
- **`sg.ClusterSize` capture.** The lvstore cluster size in bytes, queried from SPDK and cached on the ShardGroup struct: populated by `tryDiscoverExistingLvstore` on salvage / re-attach and by an explicit `BdevLvolGetLvstore` after the fresh `BdevLvolCreateLvstore`; cleared to zero on `Delete(cleanupRequired=true)`.
- **Why capture rather than hardcode.** SPDK's default when `cluster_sz=0` is 4 MiB, but `defaultClusterSize` in this repo is 1 MiB (the disk path passes 1 MiB explicitly). Hardcoding `defaultClusterSize` here would silently report ActualSize at 1/4 of the real value.

> #### How teardown works:

Reviewer: ```suggestion
#### How teardown works
```
